design-system/src/theme-provider.tsx

```diff
@@ -54,34 +57,50 @@ const validateTheme = (themeName?: string): ThemeName => {
   return 'default';
 };
 
-const ThemeProvider = (props: ThemeProviderProps) => {
-  const root = useRef<HTMLElement>(document.querySelector(':root'));
-  const localScopeElement = useRef<HTMLDivElement>(null);
-  const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
-
-  const changeTheme = useCallback((newTheme: string) => {
-    setTheme(validateTheme(newTheme));
-  }, []);
-
-  useLayoutEffect(() => {
-    const vars = toVars(themes[theme]);
-    const targetElement =
-      props.scope === 'local' ? localScopeElement.current : root.current;
-    Object.entries(vars).forEach(([key, value]) => {
-      targetElement?.style.setProperty(key, value);
-    });
-  }, [theme, props.scope]);
-
-  const value = useMemo(() => {
-    return { theme, changeTheme };
-  }, [theme, changeTheme]);
-
-  return (
-    <ThemeContext.Provider value={value}>
-      <div ref={localScopeElement}>{props.children}</div>
-    </ThemeContext.Provider>
-  );
-};
+// used to cover SSR builds (for instance in Gatsby)
+const isBrowser = typeof window !== 'undefined';
+
+const ThemeProvider = forwardRef<HTMLDivElement, ThemeProviderProps>(
+  (props, ref) => {
+    const rootRef = useRef<HTMLElement>(
+      isBrowser ? document.querySelector<HTMLElement>(':root') : null
+    );
+    const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
+
+    const changeTheme = useCallback((newTheme: string) => {
+      setTheme(validateTheme(newTheme));
+    }, []);
+
+    const localRef = ref as unknown as RefObject<HTMLDivElement>;
+
+    useEffect(() => {
+      const targetElement = localRef?.current
+        ? localRef.current
+        : rootRef?.current;
+      const vars = toVars(
+        props.customPropertiesOverrides &&
+          isObject(props.customPropertiesOverrides)
+          ? { ...themes[theme], ...props.customPropertiesOverrides }
```

Comment:
I guess we would need to update this line because the way JS merge objects by default can potentially remove props from the source theme:
```
// Example
const foo = { a: { b: 10, c: 20, d: 30 } };

const bar = { a: { d: 400 } };

const result = { ...foo, ...bar }; // { a: { d: 400 } }
// we overwrote `a` property

Perhaps we can use some helper library to help us here (example: [lodash.merge](https://lodash.com/docs/4.17.15#merge)).